### PR TITLE
[RFR] Custom save button with redirect option for Create and Edit

### LINKED ISSUE #688: [RFC] Option to redirect after item create/edit success to show page
If a resource has a show page, it may be more reasonable to redirect to show page after successful form submission.
Can we have:

an optional `afterSuccess` prop for `Create` and `Show`?
`<Create {...props} afterSuccess="show">`

or redirect to `Show` if it is available and redirect to `List` otherwise.

I can make a PR if either of them helps.

### PR DESCRIPTION
**Pull request by wesley6j**
`redirect` prop for `SaveButton`:

- [x] `handleSubmit` in `Create` and `Edit` accepts `redirect` params and creates proper action.
- [x] saga can handle actions with redirect instruction.
- [x] forms (`SimpleForm` and `TabbedForm`) accepts `toolbar` prop.
- [x] forms pass down `handleSubmit` and `onSubmit` prop to `Toolbar` and `Toolbar` passes to its buttons.
- [x] save buttons calls `onSubmit` with `redirect`.
- [x] update example to demonstrate.
- [x] test

default behavior (save button does not have redirect prop):
after create : hasEdit? go to edit ; hasShow? go to show ; go to list otherwise.
after edit:     go to list.
after delete:  list.

Notes:
1. If there are multiple buttons and `submitOnEnter` is true for the form, one of the buttons should be the "main" button (only one of the buttons is set to `submitOnEnter={true}` to be explicit).

Closes #688

### CODE REVIEW

File: src/mui/detail/Create.js [hunk 1]
```diff
@@ -15,8 +15,14 @@ class Create extends Component {
         const { location } = this.props;
         return location.pathname.split('/').slice(0, -1).join('/');
     }
-
-    handleSubmit = (record) => this.props.crudCreate(this.props.resource, record, this.getBasePath());
+    redirectTo() {
```
Comment (fzaninotto) [inline src/mui/detail/Create.js]: please rename to `getRedirectRoute`

File: src/mui/detail/Create.js [hunk 2]
```diff
@@ -15,8 +15,14 @@ class Create extends Component {
...
+        const { afterSuccess, hasShow, hasEdit } = this.props;
+        if (afterSuccess != null) return afterSuccess;
+        if (hasEdit) return 'edit';
+        if (hasShow) return 'show';
+        return 'list';
+    }
+    handleSubmit = (record) => this.props.crudCreate(this.props.resource, record, this.getBasePath(), this.redirectTo());
```
Comment (fzaninotto) [inline src/mui/detail/Create.js]: please add a newline to separate the methods

File: src/mui/detail/Edit.js [hunk 3]
```diff
@@ -43,6 +43,13 @@ export class Edit extends Component {
         return location.pathname.split('/').slice(0, -1).join('/');
     }
 
+    redirectTo() {
```
Comment (fzaninotto) [inline src/mui/detail/Edit.js]: Same as for Create, please rename to `getRedirectRoute`

**Comment by wesley6j** [comment]
@fzaninotto updated: 1. resolve conflicts, 2. keep the current defaults

**Comment by fzaninotto** [comment]
Can you rebase on `next`? Since this is a new feature, we won't merge it on `master`.

**Comment by wesley6j** [comment]
There are some problems in `next` branch that prevents a fresh new post from being displayed...

**Comment by fzaninotto** [comment]
But I think it doesn't make sense to provide the ability to override the redirection in two places (in the form AND in the submit button). There should be only one way to do it™. And the right place to allow that is in the submit button.

What I imagine is something like this:

```js
const CreateToolbar = ({invalid, submitOnEnter }) => (
    <ToolbarGroup>
        <SaveButton invalid={invalid} submitOnEnter={submitOnEnter} label="Save and add" redirect="create" />
        <SaveButton invalid={invalid} submitOnEnter={submitOnEnter} redirect="list" />
    </ToolbarGroup>
);
export const PostCreate = ({ ...props }) => (
    <Create {...props} >
        <SimpleForm toolbar={<CreateToolbar />} >
            <TextInput source="title" />
           <!--- more input components here -->
        </SimpleForm>
    </Create>
);
```

What do you think?

**Comment by wesley6j** [comment]
I totally agree! If form submit success redirect to somewhere different (like create), it is better to provide the information ("Save and add"/"Create and Show"/"Create and Edit") on the button.
I can continue with the PR to implement this.

**Comment by kodepareek** [comment]
Please consider this use case. 

I have a custom Menu to route 2 different list views List1 and List2 (with different filters and model properties displayed) to the same resource. 

More details below:
https://stackoverflow.com/questions/44138059/adding-multiple-list-views-to-same-resource-admin-on-rest/44207971#44207971

There are Edit buttons on both list views. 
On entering Edit from List2 and Saving, Edit currently redirects to to the List1 while the expectation is to redirect to List2.

With above feature we can add a string to Edit and pass it to a custom location. However it would be great to have a way to also go back to the origination URL i.e go back 1 step in history, to the originating list view. 

**Suggestion** - we add a prop `origin` to the EditButton component. This is passed to Edit. Allowing us to check the origin List view and set the redirect string appropriately.

**Comment by kodepareek** [comment]
@wesley6j  would be really great if you could also please consider the use case in my previous comment

**Comment by wesley6j** [comment]
@kodepareek Custom `EditButton` is not in this PR's scope (And I think you need custom `CreateButton`, `ShowButton`, `ListButton` and `DeleteButton` as well). But this PR can help you with `SaveButton`. 

If you put `status=NEW` in all relevant urls' parameters, you can find out where to link to in a button by checking current url. This is the same for `SaveButton`, edit this [line](https://github.com/wesley6j/admin-on-rest/blob/ca64b6add1d432ed9ccbb1d6bb455b7d1cee4b15/src/mui/button/SaveButton.js#L20) you can replace redirect with your url calculation result.

**Comment by kodepareek** [comment]
@wesley6j  not very comfortable with tweaking the framework. Plus will lead to long term maintenance issues. The SaveButton is tied into Edit component and overriding it would mean tinkering with Edit. 

I would then suggest that if not custom URLs then can we make a provision to redirect to the origin URL. i.e go back one step in history. This would solve the use case of custom Menus (where multiple views for the same resource.)

**Comment by edorivai** [comment]
Hey, as per #733, I would like the possibility to stay on the current page after saving. Would that be possible with the approach in this PR?

**Comment by edorivai** [comment]
Basically, for me something like this would suffice:

```js
<SaveButton redirect={false} />
```

But I think this would not work currently, due to `Edit` falling back to `this.getRedirectRoute()` if `redirect` is falsy:

```js
this.props.crudUpdate(this.props.resource, this.props.id, record, this.props.data, this.getBasePath(), redirect || this.getRedirectRoute());
```

**Comment by wesley6j** [comment]
@edorivai Thanks. I did not notice this problem.

**Comment by edorivai** [comment]
@wesley6j Cool! So you'll take that use case into account in this PR? If so, can close #733.

**Comment by wesley6j** [comment]
@edorivai There is one small problem though. In case of `redirect={false}`, this button keeps spinning after form is submitted because currently only `LOCATION_CHANGE` or submit failure can set `state.admin.saving` to false.
`

**Comment by edorivai** [comment]
Hmm, any way I can jump in and help with that issue?

**Comment by wesley6j** [comment]
@edorivai I wonder if `redirect="create"` or `redirect="edit"` works for you. I think it might be a bad idea that redirect can be either boolean or string.

**Comment by edorivai** [comment]
Hmm, I don't think that would be a very good UX for us, since we are using a `TabbedForm`, and AFAIK it doesn't add the current tab to the URL, so I'm assuming the user might be sent to a different tab upon saving.

**Comment by fzaninotto** [comment]
Almost ready. I just don't understand the purpose of the two `handleSubmit` and `onSubmit` props in `<SaveButton>`, it seems that you could do with just one. Is it the rest of one of your earlier attempts?

**Comment by wesley6j** [comment]
The two functions are different:
`handleSubmit` is a redux-form function.
`onSubmit` is the function created in `Create` or `Edit` that actually uses `crudUpdate` to submit data to server.
The two functions are used to compose this:
```jsx
handleSubmit(values => onSubmit(values, redirect))();
```

There is a similar [example](https://stackoverflow.com/a/37173842/2422573) to explain the above.

**Comment by fzaninotto** [comment]
I've spent the last 15 minutes figuring out this onSubmit / handleSubmit thing, but I'm sure I'll have to spend even more next time I need to work on it. As is, the code isn't maintainable enough. Props being passed three levels down, methods named the same but not doing the same, it's all too confusing. Could you try and refactor it for better clarity, maybe by using redux form actions instead of props ?

**Comment by wesley6j** [comment]
@fzaninotto What about composing the two functions at a earlier stage (maybe inside `SimpleForm` and `TabbedForm`) to a single `submitWithRedirect` function? But I still have to pass it down multiple layers...

**Comment by fzaninotto** [comment]
That sounds like a reasonable track to follow. Test it, don't hesitate to rename props if that makes sense. If the code is still not clear enough after that change, I'll ask you for even more changes ;)

File: src/mui/button/SaveButton.js [hunk 4]
```diff
@@ -60,19 +60,17 @@ SaveButton.propTypes = {
...
     redirect: PropTypes.oneOfType([
         PropTypes.string,
         PropTypes.bool,
     ]),
 };
 
 SaveButton.defaultProps = {
-    handleSubmit: submit => (values, redirect) => { submit(values, redirect); },
-    onSubmit: () => {},
+    submitWithRedirect: () => {},
```
Comment (djhi) [inline src/mui/button/SaveButton.js]: should be named `onSubmitWithRedirect`

Comment (wesley6j) [inline src/mui/button/SaveButton.js]: Thanks. Updated.

**Comment by wesley6j** [comment]
🍻 🍻 🍻

**Pull request merged**